swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -1292,7 +1403,111 @@ string ppReprType(DataFlowType t) { none() }
  * a node of type `t1` to a node of type `t2`.
  */
 pragma[inline]
-predicate compatibleTypes(DataFlowType t1, DataFlowType t2) { any() }
+predicate compatibleTypes(DataFlowType t1, DataFlowType t2) {
+  subType(t1) = subType(t2)
+  or
+  exists(BoundGenericType bound1, BoundGenericType bound2 |
+    stripType(t1.asType()).(BoundGenericType) = bound1 and
+    stripType(t2.asType()).(BoundGenericType) = bound2 and
+    bound1.getDeclaration() = bound2.getDeclaration() and
+    forall(int index | index in [0 .. bound1.getNumberOfArgTypes() - 1] |
+      isAnyOrUnboundType(bound1.getArgType(index).getCanonicalType()) or
+      isAnyOrUnboundType(bound2.getArgType(index).getCanonicalType())
+    )
+  )
```

Comment:
As I said on Slack, this may perform better if the `forall` is rewritten to explicit recursion. But if this doesn't blow up on anything we're aware of then this LGTM!